tests/csapi/apidoc_login_test.go

```diff
@@ -133,5 +134,30 @@ func TestLogin(t *testing.T) {
 				},
 			})
 		})
+
+		// Regression test for https://github.com/matrix-org/dendrite/issues/2287
+		t.Run("Login with uppercase username works and syncing afterwards also", func(t *testing.T) {
+			t.Parallel()
+			// login should be possible with uppercase username
+			res := unauthedClient.MustDoFunc(t, "POST", []string{"_matrix", "client", "r0", "login"}, client.WithRawBody(json.RawMessage(`{
+				"type": "m.login.password",
+				"identifier": {
+					"type": "m.id.user",
+					"user": "@Test_login_user:hs1"
+				},
+				"password": "superuser"
+			}`)))
+			// extract access_token
+			body, err := ioutil.ReadAll(res.Body)
```

Comment:
You can use [must.ParseJSON](https://pkg.go.dev/github.com/matrix-org/complement@v0.0.0-20211217181955-5742fddd409d/internal/must#ParseJSON) with `res.Body` here.